lib/flight_job/models/template.rb

```diff
@@ -67,25 +68,22 @@ class Template < ApplicationModel
       "$defs" => {}.merge!(SchemaDefs::VALIDATOR_DEF).merge!(SchemaDefs::QUESTION_DEF)
     })
 
-    def self.load_all
-      templates = Dir.glob(new(id: '*').metadata_path).map do |path|
+    def self.load_all(opts = nil)
+      Dir.glob(new(id: '*').metadata_path).map do |path|
         id = File.basename(File.dirname(path))
-        new(id: id)
-      end
-
-      templates.each do |template|
-        next if template.valid?
-        FlightJob.logger.warn("Invalid template detected upon load: #{template.id}")
-        FlightJob.logger.warn(template.errors.full_messages.join("\n"))
-      end
-
-      templates.sort!
-
-      templates.each_with_index do |t, idx|
-        t.index = idx + 1
-      end
-
-      templates
+        template = new(id: id)
+        if template.pass_filter?(opts)
+          if template.valid?
+            template
+          else
+            FlightJob.logger.warn("Invalid template detected upon load: #{template.id}")
+            FlightJob.logger.warn(template.errors.full_messages.join("\n"))
+            nil
+          end
+        end
+      end.compact
+         .sort
+         .each_with_index { |t, idx| t.index = idx + 1 }
```

Comment:
There has been a change of behavior introduced here.  Previously, a template was assigned an index and included in the display of `list-templates` even if it was invalid.

This is important to template authors as otherwise they can come to the conclusion that flight job knows nothing of their template.

```sh
[vagrant@chead1 flight-job]$ /opt/flight/bin/flight job list-templates
┌───────┬─────────────────────────┬─────────────────────────────────────────────┐
│ Index │ ID                      │ Name                                        │
├───────┼─────────────────────────┼─────────────────────────────────────────────┤
│ 1     │ desktop-on-login-node   │ Interactive desktop session on login node   │
│ 2     │ desktop-on-compute-node │ Interactive desktop session on compute node │
│ 3     │ simple*                 │ (none)                                      │
│ 4     │ simple-array            │ Simple serial array                         │
│ 5     │ mpi-slots               │ Multiple slot                               │
│ 6     │ mpi-nodes               │ Multiple node                               │
│ 7     │ smp                     │ SMP multiple slot                           │
└───────┴─────────────────────────┴─────────────────────────────────────────────┘
 * Invalid template
```

With these changes, the invalid template is no longer displayed and its index is not "taken".

```sh
[vagrant@chead1 flight-job]$ /opt/flight/bin/bundle exec ./bin/job list-templates                                  
┌───────┬─────────────────────────┬─────────────────────────────────────────────┐
│ Index │ ID                      │ Name                                        │
├───────┼─────────────────────────┼─────────────────────────────────────────────┤
│ 1     │ desktop-on-login-node   │ Interactive desktop session on login node   │
│ 2     │ desktop-on-compute-node │ Interactive desktop session on compute node │
│ 3     │ simple-array            │ Simple serial array                         │
│ 4     │ mpi-slots               │ Multiple slot                               │
│ 5     │ mpi-nodes               │ Multiple node                               │
│ 6     │ smp                     │ SMP multiple slot                           │
└───────┴─────────────────────────┴─────────────────────────────────────────────┘
```

---

The template's `Index` should be stable in the face of invalid templates and also when some templates are excluded as a result of filtering.  This is because the template's index can be used as input to the `info-template`, `create-script` and other commands.  In other words, adding or removing templates can adjust the indexes, but editing a template should not, and filtering templates should not.

E.g., here the user filters the templates with the intention of getting info about the `mpi-slots` template.  However, they use the template's index and obtain info about the `desktop-on-login-node` template instead.

```sh
[vagrant@chead1 flight-job]$ /opt/flight/bin/bundle exec ./bin/job list-templates --id '*mpi*'
┌───────┬───────────┬───────────────┐
│ Index │ ID        │ Name          │
├───────┼───────────┼───────────────┤
│ 1     │ mpi-slots │ Multiple slot │
│ 2     │ mpi-nodes │ Multiple node │
└───────┴───────────┴───────────────┘
[vagrant@chead1 flight-job]$ /opt/flight/bin/bundle exec ./bin/job info-template 1                                 
      ID: desktop-on-login-node
    Name: Interactive desktop session on login node
Synopsis: Start an interactive desktop session on a login node and submit a job to run on a compute node.

DESCRIPTION:
An interactive desktop session will start on a login node.  Once that
session is running, resources will be requested from the scheduler to run
the job script.  If the job script starts a graphical application it will be
available from inside the desktop session.
[vagrant@chead1 flight-job]$ 
```